### PR TITLE
CTCTRADERS-576

### PR DESCRIPTION
**Pull request by kenneth-rayner**
Update for unloading remarks rejection

### CODE REVIEW

File: app/connectors/UnloadingConnectorImpl.scala [hunk 1]
```diff
@@ -53,9 +57,33 @@ class UnloadingConnectorImpl @Inject()(val config: FrontendAppConfig, val http:
...
+    val serviceUrl: String = s"${config.arrivalsBackend}/movements/arrivals/${arrivalId.value}/messages/summary"
+    http.GET[HttpResponse](serviceUrl) map {
+      case responseMessage if is2xx(responseMessage.status) =>
+        Some(responseMessage.json.as[MessagesSummary])
+      case _ => None
+    }
+  }
+
+  def getRejectionMessage(rejectionLocation: String)(implicit hc: HeaderCarrier): Future[Option[UnloadingRemarksRejectionMessage]] = {
+    val serviceUrl = s"${config.arrivalsBackendBaseUrl}$rejectionLocation"
```
Comment (chris-ianson) [inline app/connectors/UnloadingConnectorImpl.scala:71]: Should the rejectionLocation not contain the full uri of the resource? We expose the getSummary endpoints for consumers to pull back data? Why don't we return the full uri within the json?

File: app/models/ErrorType.scala [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  case object UnknownMrn extends MRNError(90)
+  case object DuplicateMrn extends MRNError(91)
+  case object InvalidMrn extends MRNError(93)
+
+  val mrnValues = Seq(
+    UnknownMrn,
+    DuplicateMrn,
+    InvalidMrn
+  )
```
Comment (chris-ianson) [inline app/models/ErrorType.scala:53]: Remove these, don't accept MRN within an IE044. We send it through but it's taken from IE043 (we don't have any control over this data - if we get this kind of rejection back there's been a serious error outside of this service)

File: app/models/ErrorType.scala [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import scala.xml.NodeSeq
+
+sealed trait ErrorType {
+  val code: Int
+}
+
+object ErrorType extends Enumerable.Implicits {
+
+  sealed abstract class GenericError(val code: Int) extends ErrorType
+  sealed abstract class MRNError(val code: Int) extends ErrorType
```
Comment (chris-ianson) [inline app/models/ErrorType.scala:31]: Remove MRNError, don't support it

File: conf/app.routes [hunk 4]
```diff
@@ -86,3 +86,5 @@ GET        /:arrivalId/total-number-of-packages                  controllers.Tot
 POST       /:arrivalId/total-number-of-packages                    controllers.TotalNumberOfPackagesController.onSubmit(arrivalId: ArrivalId, mode: Mode = NormalMode)
 GET        /:arrivalId/change-total-number-of-packages                          controllers.TotalNumberOfPackagesController.onPageLoad(arrivalId: ArrivalId, mode: Mode = CheckMode)
 POST       /:arrivalId/change-total-number-of-packages                          controllers.TotalNumberOfPackagesController.onSubmit(arrivalId: ArrivalId, mode: Mode = CheckMode)
+
+GET        /:arrivalId/unloading-remarks-rejection                       controllers.UnloadingRemarksRejectionController.onPageLoad(arrivalId: ArrivalId)
```
Comment (chris-ianson) [inline conf/app.routes:90]: Do you think _rejection_ is all we need here? We can only have an unloading rejection and the unloading url indicates what journey a user is in. What do people think?

File: app/models/MessagesLocation.scala [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+package models
+
+import play.api.libs.json.{__, Reads}
+
+case class MessagesLocation(unloadingRemarks: String, unloadingRemarksRejection: Option[String] = None)
+
+object MessagesLocation {
+
+  import play.api.libs.functional.syntax._
+  // TODO: update Message codes
```
[the remark this answers is not in the document]
Comment (kenneth-rayner) [reply app/models/MessagesLocation.scala:26]: Done

File: app/models/MessagesSummary.scala [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models
+
+import play.api.libs.json.{__, Reads}
+//TODO Add tests
```
[the remark this answers is not in the document]
Comment (kenneth-rayner) [reply app/models/MessagesSummary.scala:20]: Done

File: app/connectors/UnloadingConnectorImpl.scala (hunk 1 above)
Comment (kenneth-rayner) [reply app/connectors/UnloadingConnectorImpl.scala:71]: Rejection location holds full URL, here we are just adding localhost and port number

File: conf/app.routes (hunk 4 above)
Comment (kenneth-rayner) [reply conf/app.routes:90]: That will be looked at by Fred

File: app/connectors/UnloadingConnectorImpl.scala (hunk 1 above)
Comment (chris-ianson) [reply app/connectors/UnloadingConnectorImpl.scala:77]: rejectionLocation is not the full url. arrivalsBackendBaseUrl includes _transit-movements-trader-at-destination_ and then we tag on rejectionLocation. I'm just wondering if _transit-movements-trader-at-destination_ part should be returned in the getSummary call to the backend? @ashwinirudrappa What are your thoughts on this?

File: app/viewModels/UnloadingRemarksRejectionViewModel.scala [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import models.{ArrivalId, CheckMode}
+import play.api.i18n.Messages
+import play.api.libs.json.{JsObject, Json}
+import uk.gov.hmrc.viewmodels.SummaryList.{Action, Key, Row, Value}
+import uk.gov.hmrc.viewmodels._
+import viewModels.sections.Section
+
+case class UnloadingRemarksRejectionViewModel(page: String, json: JsObject)
+
+object UnloadingRemarksRejectionViewModel {
```
Comment (chris-ianson) [inline app/viewModels/UnloadingRemarksRejectionViewModel.scala:28]: Can we add a spec for this please? (see the other view model specs)

File: conf/application-json-logger.xml [hunk 8]
```diff
@@ -6,6 +6,7 @@
     </appender>
 
     <logger name="uk.gov" level="WARN"/>
+    <logger name="controllers.UnloadingRemarksRejectionController" level="${Log.UnloadingRemarksRejectionController:-DEBUG}"/>
```
Comment (chris-ianson) [inline conf/application-json-logger.xml:9]: Is this setting UnloadingRemarksRejectionController to always log DEBUG outputs?

Comment (kenneth-rayner) [reply conf/application-json-logger.xml:9]: Yes

File: it/connectors/UnloadingConnectorSpec.scala [hunk 9]
```diff
@@ -127,6 +137,107 @@ class UnloadingConnectorSpec extends FreeSpec with ScalaFutures with
...
+      }
+    }
+
+    "getRejectionMessage" - {
+      "must return valid 'rejection message'" in {
+        val genRejectionError     = arbitrary[ErrorType].sample.value
+        val rejectionXml: NodeSeq =             <CC058A>
+          <HEAHEA>
+            <DocNumHEA5>19IT021300100075E9</DocNumHEA5>
+            <UnlRemRejDatHEA218>20191018</UnlRemRejDatHEA218>
```
Comment (chris-ianson) [inline it/connectors/UnloadingConnectorSpec.scala:184]: Would be good to generate all values here. But at a minimum could we at least generate the LocalDate value we assert against (you can use _Format.dateFormatted([localdatevalue])_ to format it for the UnlRemRejDatHEA218 node)

File: test/controllers/UnloadingRemarksRejectionControllerSpec.scala [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  private val mockUnloadingRemarksRejectionService = mock[UnloadingRemarksRejectionService]
+
+  override def beforeEach: Unit = {
+    super.beforeEach
+    reset(mockUnloadingRemarksRejectionService)
+  }
+
+  "UnloadingRemarksRejection Controller" - {
+
+    "return OK and the Rejection view for a GET when unloading rejection message returns a Some" in {
```
Comment (chris-ianson) [inline test/controllers/UnloadingRemarksRejectionControllerSpec.scala:58]: Could we add a test which returns TechnicalDifficulties when the error length is greater than 1

File: conf/application-json-logger.xml (hunk 8 above)
Comment (chris-ianson) [reply conf/application-json-logger.xml:9]: 👍

File: it/connectors/UnloadingConnectorSpec.scala (hunk 9 above)
Comment (kenneth-rayner) [reply it/connectors/UnloadingConnectorSpec.scala:184]: We don't format it when we read an Xml

**Pull request closed without merging**